include/rsl/monad.hpp

```diff
@@ -0,0 +1,136 @@
+#pragma once
+
+#include <tl_expected/expected.hpp>
+
+#include <optional>
+
+namespace rsl {
+
+/**
+ * @brief      Monad optional bind
+ *
+ * @param[in]  opt   The input optional
+ * @param[in]  f     The function
+ *
+ * @tparam     T     The input type
+ * @tparam     F     The function
+ *
+ * @return     Return type of f
+ */
+template <typename T, typename F>
+constexpr auto mbind(std::optional<T> const& opt, F f) -> decltype(f(opt.value())) {
```

Comment:
Is this `decltype` to ensure that any CV-qualifications on `f`'s return type are preserved?

Reply:
decltype lets you put expressions into it like function calls and it will evaluate the type. I did this to deduce the return type of the f function so it doesn't have to be specified as a template parameter.

I don't think this has anything to do with CV-qualifications, and more to do with just being a work-around to get the return type of a function that is passed in as a template parameter.